pkg/lint/rules/dependencies.go

```diff
@@ -0,0 +1,82 @@
+/*
+Copyright The Helm Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package rules // import "helm.sh/helm/v3/pkg/lint/rules"
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/pkg/errors"
+
+	"helm.sh/helm/v3/pkg/chart"
+	"helm.sh/helm/v3/pkg/chart/loader"
+	"helm.sh/helm/v3/pkg/lint/support"
+)
+
+// Dependencies runs lints against a chart's dependencies
+//
+// See https://github.com/helm/helm/issues/7910
+func Dependencies(linter *support.Linter) {
+	c, err := loader.LoadDir(linter.ChartDir)
+	if !linter.RunLinterRule(support.ErrorSev, "", validateChartFormat(err)) {
+		return
+	}
+
+	linter.RunLinterRule(support.ErrorSev, linter.ChartDir, validateDependencyInMetadata(c))
+	linter.RunLinterRule(support.WarningSev, linter.ChartDir, validateDependencyInChartsDir(c))
```

Comment:
There are many cases where someone would have a dependency listed in a chart but not have it in the `charts` directory. Should this be a warning?

Reply:
The issue is that there are never _supposed_ to be cases like this. Helm 3 assumes (for proper behavior) that the contents of `Chart.Metadata.Dependencies` is the same as `Chart.Dependencies()`. When these are out of sync, weird things happen.

So we have a case where the code assumes something, but never explicitly documented this assumption. However, many features of Helm need this to be the case.

My suspicion is that a lot of the issues filed about problems with deeply nested charts will simply vanish when people are forced to keep their stated dependencies in sync with the implicit dependencies they drop in `charts/`.